torch/csrc/distributed/c10d/Hooks.cpp

```diff
@@ -15,44 +15,22 @@ namespace c10d {
 
 namespace {
 
-std::atomic<bool> event_queue_enabled = false;
-int sync_pipe;
-std::mutex event_queue_lock;
-std::deque<details::EventInfo> event_queue;
-
+std::mutex callbacks_lock;
+std::vector<CollectiveEventCallback> callbacks_list;
 } // namespace
 
-void enable_event_collection(int pipe) {
-  sync_pipe = pipe;
-  event_queue_enabled.store(true);
+TORCH_API void register_collective_callback(
+    CollectiveEventCallback&& callback) {
+  std::unique_lock<std::mutex> lock(callbacks_lock);
+  callbacks_list.push_back(std::move(callback));
 }
 
 namespace details {
 
-// we start dropping events after this
-const size_t MAX_QUEUE_SIZE = 512;
-
-bool dequeue_c10d_event(EventInfo& evt) {
-  std::unique_lock<std::mutex> lock(event_queue_lock);
-  if (event_queue.size() == 0) {
-    return false;
-  }
-  evt = event_queue.front();
-  event_queue.pop_front();
-  return true;
-}
-
 void enqueue_c10d_event(EventInfo&& evt) {
-  if (!event_queue_enabled.load())
-    return;
-
-  std::unique_lock<std::mutex> lock(event_queue_lock);
-  if (event_queue.size() >= MAX_QUEUE_SIZE) {
-    event_queue.back().drop_count++;
-  } else {
-    event_queue.push_back(std::move(evt));
-    char m = 'x';
-    write(sync_pipe, &m, 1);
+  std::unique_lock<std::mutex> lock(callbacks_lock);
+  for (auto& cb : callbacks_list) {
+    cb(evt);
```

Comment:
I am not sure if I am following why do we want to call the callback when we insert event into the queue?

Reply:
this PR refactors so that there is no queue by default.

i think it would be clearer to just rename `enqueue_c10d_event` to `call_c10d_event_callbacks`, so ill make that change.

A queue is a possible implementation detail of a callback.  The python system uses a queue.  If another customer implemented callbacks they may choose to use a queue but they control it.

The requirement is the callback(s) that are registered need to act quickly and safely, delegating their real work to some other thread or process.

Reply:
yea, in this case 

A) the python callback system uses its own python thread to 'drain' the python event queue, and all the c++ callback does is push events onto the python event queue.  Most of this implementation is in the previous PR, but it gets refactored a bit in this PR to make the c++ callback mechanism more generic

B) any other c++ callback that some user registers is expected to use their own queue and background thread if they need to do long-running things.